Jupyter QtConsole 5.0.3
Python 3.8.8 (default, Apr 13 2021, 15:08:03) [MSC v.1916 64 bit (AMD64)]
Type 'copyright', 'credits' or 'license' for more information
IPython 7.22.0 -- An enhanced Interactive Python. Type '?' for help.

import sqlite3

connection = sqlite3.connect('books.db')

import pandas as pd

pd.options.display.max_columns = 10

pd.read_sql('SELECT * FROM authors', connection,
            index_col=['id'])
---------------------------------------------------------------------------
OperationalError                          Traceback (most recent call last)
~\anaconda3\lib\site-packages\pandas\io\sql.py in execute(self, *args, **kwargs)
   1724         try:
-> 1725             cur.execute(*args, **kwargs)
   1726             return cur

OperationalError: no such table: authors

The above exception was the direct cause of the following exception:

DatabaseError                             Traceback (most recent call last)
<ipython-input-5-e20235c08b3f> in <module>
----> 1 pd.read_sql('SELECT * FROM authors', connection,
      2             index_col=['id'])

~\anaconda3\lib\site-packages\pandas\io\sql.py in read_sql(sql, con, index_col, coerce_float, params, parse_dates, columns, chunksize)
    493 
    494     if isinstance(pandas_sql, SQLiteDatabase):
--> 495         return pandas_sql.read_query(
    496             sql,
    497             index_col=index_col,

~\anaconda3\lib\site-packages\pandas\io\sql.py in read_query(self, sql, index_col, coerce_float, params, parse_dates, chunksize)
   1769 
   1770         args = _convert_params(sql, params)
-> 1771         cursor = self.execute(*args)
   1772         columns = [col_desc[0] for col_desc in cursor.description]
   1773 

~\anaconda3\lib\site-packages\pandas\io\sql.py in execute(self, *args, **kwargs)
   1735 
   1736             ex = DatabaseError(f"Execution failed on sql '{args[0]}': {exc}")
-> 1737             raise ex from exc
   1738 
   1739     @staticmethod

DatabaseError: Execution failed on sql 'SELECT * FROM authors': no such table: authors

"C:\Users\Norma\OneDrive\Documents\ch17\books.db""C:\Users\Norma\OneDrive\Documents\ch17\books.db"
  File "<ipython-input-6-40fcdc0f0eb8>", line 1
    "C:\Users\Norma\OneDrive\Documents\ch17\books.db""C:\Users\Norma\OneDrive\Documents\ch17\books.db"
    ^
SyntaxError: (unicode error) 'unicodeescape' codec can't decode bytes in position 2-3: truncated \UXXXXXXXX escape


connection = sqlite.connect("C:\\Users\\Norma\\OneDrive\\Documents\\ch17\\books.db")
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-7-d66cf51021ad> in <module>
----> 1 connection = sqlite.connect("C:\\Users\\Norma\\OneDrive\\Documents\\ch17\\books.db")

NameError: name 'sqlite' is not defined

connection = sqlite3.connect("C:\\Users\\Norma\\OneDrive\\Documents\\ch17\\books.db")

pd.read_sql('SELECT * FROM authors', connection,
            index_col=['id'])
Out[9]: 
        first    last
id                   
1        Paul  Deitel
2      Harvey  Deitel
3       Abbey  Deitel
4         Dan   Quirk
5   Alexander    Wald

pd.read_sql('SELECT * FROM titles', connection)
Out[10]: 
         isbn                             title  edition copyright
0  0135404673     Intro to Python for CS and DS        1      2020
1  0132151006     Internet & WWW How to Program        5      2012
2  0134743350               Java How to Program       11      2018
3  0133976890                  C How to Program        8      2016
4  0133406954  Visual Basic 2012 How to Program        6      2014
5  0134601548          Visual C# How to Program        6      2017
6  0136151574         Visual C++ How to Program        2      2008
7  0134448235                C++ How to Program       10      2017
8  0134444302            Android How to Program        3      2017
9  0134289366         Android 6 for Programmers        3      2016

df = pd.read_sql('SELECT * FROM author_ISBN', connection)

df.head()
Out[12]: 
   id        isbn
0   1  0134289366
1   2  0134289366
2   5  0134289366
3   1  0135404673
4   2  0135404673

pd.read_sql('SELECT first, last FROM authors', connection)
Out[13]: 
       first    last
0       Paul  Deitel
1     Harvey  Deitel
2      Abbey  Deitel
3        Dan   Quirk
4  Alexander    Wald

pd.read_sql("""SELECT title, edition, copyright
                FROM titles
                WHERE copyright > '2016""", connection)
---------------------------------------------------------------------------
OperationalError                          Traceback (most recent call last)
~\anaconda3\lib\site-packages\pandas\io\sql.py in execute(self, *args, **kwargs)
   1724         try:
-> 1725             cur.execute(*args, **kwargs)
   1726             return cur

OperationalError: unrecognized token: "'2016"

The above exception was the direct cause of the following exception:

DatabaseError                             Traceback (most recent call last)
<ipython-input-14-725abdb02ca1> in <module>
----> 1 pd.read_sql("""SELECT title, edition, copyright
      2                 FROM titles
      3                 WHERE copyright > '2016""", connection)

~\anaconda3\lib\site-packages\pandas\io\sql.py in read_sql(sql, con, index_col, coerce_float, params, parse_dates, columns, chunksize)
    493 
    494     if isinstance(pandas_sql, SQLiteDatabase):
--> 495         return pandas_sql.read_query(
    496             sql,
    497             index_col=index_col,

~\anaconda3\lib\site-packages\pandas\io\sql.py in read_query(self, sql, index_col, coerce_float, params, parse_dates, chunksize)
   1769 
   1770         args = _convert_params(sql, params)
-> 1771         cursor = self.execute(*args)
   1772         columns = [col_desc[0] for col_desc in cursor.description]
   1773 

~\anaconda3\lib\site-packages\pandas\io\sql.py in execute(self, *args, **kwargs)
   1735 
   1736             ex = DatabaseError(f"Execution failed on sql '{args[0]}': {exc}")
-> 1737             raise ex from exc
   1738 
   1739     @staticmethod

DatabaseError: Execution failed on sql 'SELECT title, edition, copyright
                FROM titles
                WHERE copyright > '2016': unrecognized token: "'2016"

pd.read_sql("""SELECT title, edition, copyright
                FROM titles
                WHERE copyright > '2016'""", connection)
Out[15]: 
                           title  edition copyright
0  Intro to Python for CS and DS        1      2020
1            Java How to Program       11      2018
2       Visual C# How to Program        6      2017
3             C++ How to Program       10      2017
4         Android How to Program        3      2017

pd.read_sql("""SELECT id, first, last
                FROM authors
                WHERE last LIKE 'D%'""",
                connection, index_col['id'])
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-16-68630a537271> in <module>
      2                 FROM authors
      3                 WHERE last LIKE 'D%'""",
----> 4                 connection, index_col['id'])

NameError: name 'index_col' is not defined

pd.read_sql("""SELECT id, first, last
                FROM authors
                WHERE last LIKE 'D%'""",
                connection, index_col=['id'])
Out[17]: 
     first    last
id                
1     Paul  Deitel
2   Harvey  Deitel
3    Abbey  Deitel

pd.read_sql("""SELECT id, first, last
                FROM authors
                WHERE last LIKE '_b%'""",
                connection, index_col=['id'])
Out[18]: 
Empty DataFrame
Columns: [first, last]
Index: []

pd.read_sql("""SELECT id, first, last
                FROM authors
                WHERE first LIKE '_b%'""",
                connection, index_col=['id'])
Out[19]: 
    first    last
id               
3   Abbey  Deitel

pd.read_sql('SELECT title FROM titles ORDER BY title ASC',
            connection)
Out[20]: 
                              title
0         Android 6 for Programmers
1            Android How to Program
2                  C How to Program
3                C++ How to Program
4     Internet & WWW How to Program
5     Intro to Python for CS and DS
6               Java How to Program
7  Visual Basic 2012 How to Program
8          Visual C# How to Program
9         Visual C++ How to Program

pd.read_sql("""SELECT id, first last
            FROM authors
            ORDER BY last, first""",
            connection, index_col = ['id'])            
Out[21]: 
         last
id           
3       Abbey
5   Alexander
4         Dan
2      Harvey
1        Paul

pd.read_sql("""SELECT id, first last
            FROM authors
            ORDER BY last DESC, first ASC""",
            connection, index_col = ['id'])
Out[22]: 
         last
id           
1        Paul
2      Harvey
4         Dan
5   Alexander
3       Abbey

pd.read_sql("""SELECT id, first, last
            FROM authors
            ORDER BY last DESC, first ASC""",
            connection, index_col = ['id'])
Out[23]: 
        first    last
id                   
5   Alexander    Wald
4         Dan   Quirk
3       Abbey  Deitel
2      Harvey  Deitel
1        Paul  Deitel

pd.read_sql("""SELECT id, first, last
            FROM authors
            ORDER BY last, first""",
            connection, index_col = ['id'])
Out[24]: 
        first    last
id                   
3       Abbey  Deitel
2      Harvey  Deitel
1        Paul  Deitel
4         Dan   Quirk
5   Alexander    Wald

pd.read_sql("""SELECT isbn, title, edition, copyright
            FROM titles
            WHERE title LIKE '%How to Program'
            ORDER BY title""", connection)
Out[25]: 
         isbn                             title  edition copyright
0  0134444302            Android How to Program        3      2017
1  0133976890                  C How to Program        8      2016
2  0134448235                C++ How to Program       10      2017
3  0132151006     Internet & WWW How to Program        5      2012
4  0134743350               Java How to Program       11      2018
5  0133406954  Visual Basic 2012 How to Program        6      2014
6  0134601548          Visual C# How to Program        6      2017
7  0136151574         Visual C++ How to Program        2      2008

pd.read_sql("""SELECT first, last, isbn
            FROM authors
            INNER JOIN author_ISBN
                ON authors.id = author_ISBN.id
            ORDER BY last, first""", connection).head()
Out[26]: 
    first    last        isbn
0   Abbey  Deitel  0132151006
1   Abbey  Deitel  0133406954
2  Harvey  Deitel  0134289366
3  Harvey  Deitel  0135404673
4  Harvey  Deitel  0132151006

cursor = connection.cursor()

cursor = connection.cursor.execute("""INSERT INTO authors (first, last)
                                    VALUES ('Sue', 'Red')""")
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-28-c4eb2fdff597> in <module>
----> 1 cursor = connection.cursor.execute("""INSERT INTO authors (first, last)
      2                                     VALUES ('Sue', 'Red')""")

AttributeError: 'builtin_function_or_method' object has no attribute 'execute'

cursor = cursor.execute("""INSERT INTO authors (first, last)
                                    VALUES ('Sue', 'Red')""")

pd.read('SELECT id, first, last FROM authors',
        connection, index_col=['id'])
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-30-f7f16c73ec1d> in <module>
----> 1 pd.read('SELECT id, first, last FROM authors',
      2         connection, index_col=['id'])

~\anaconda3\lib\site-packages\pandas\__init__.py in __getattr__(name)
    242         return _SparseArray
    243 
--> 244     raise AttributeError(f"module 'pandas' has no attribute '{name}'")
    245 
    246 

AttributeError: module 'pandas' has no attribute 'read'

pd.read_sql('SELECT id, first, last FROM authors',
        connection, index_col=['id'])
Out[31]: 
        first    last
id                   
1        Paul  Deitel
2      Harvey  Deitel
3       Abbey  Deitel
4         Dan   Quirk
5   Alexander    Wald
6         Sue     Red

cursor = cursor.execute("""UPDATE authors SET last = 'Black'
                        WHERE last = 'Red AND first = 'Sue'""")
---------------------------------------------------------------------------
OperationalError                          Traceback (most recent call last)
<ipython-input-32-428d6191c8fd> in <module>
----> 1 cursor = cursor.execute("""UPDATE authors SET last = 'Black'
      2                         WHERE last = 'Red AND first = 'Sue'""")

OperationalError: near "Sue": syntax error

cursor = cursor.execute("""UPDATE authors SET last = 'Black'
                        WHERE last = 'Red' AND first = 'Sue'""")

cursor.rowcount
Out[34]: 1

pd.read_sql('SELECT id, first, last FROM authors',
        connection, index_col=['id'])
Out[35]: 
        first    last
id                   
1        Paul  Deitel
2      Harvey  Deitel
3       Abbey  Deitel
4         Dan   Quirk
5   Alexander    Wald
6         Sue   Black

cursor = cursor.execute('DELETE FROM authors WHERE id =6')

cursor.rowcount
Out[37]: 1

pd.read_sql('SELECT id, first, last FROM authors',
        connection, index_col=['id'])
Out[38]: 
        first    last
id                   
1        Paul  Deitel
2      Harvey  Deitel
3       Abbey  Deitel
4         Dan   Quirk
5   Alexander    Wald

pd.read_sql("""SELECT title, edition FROM titles
                ORDER BY edition DESC""", connection)head(3)
  File "<ipython-input-39-39339a6e7212>", line 2
    ORDER BY edition DESC""", connection)head(3)
                                         ^
SyntaxError: invalid syntax


pd.read_sql("""SELECT title, edition FROM titles
                ORDER BY edition DESC""", connection).head(3)
Out[40]: 
                 title  edition
0  Java How to Program       11
1   C++ How to Program       10
2     C How to Program        8

pd.read_sql("""SELECT * FROM authors
                WHERE first LIKE 'A%'""", connection)
Out[41]: 
   id      first    last
0   3      Abbey  Deitel
1   5  Alexander    Wald

pd.read_sql("""SELECT isbn, title, edition, copyright
                FROM titles
                WHERE title NOT LIKE '%How to Program'
                ORDER BY titel""", connection)
---------------------------------------------------------------------------
OperationalError                          Traceback (most recent call last)
~\anaconda3\lib\site-packages\pandas\io\sql.py in execute(self, *args, **kwargs)
   1724         try:
-> 1725             cur.execute(*args, **kwargs)
   1726             return cur

OperationalError: no such column: titel

The above exception was the direct cause of the following exception:

DatabaseError                             Traceback (most recent call last)
<ipython-input-42-26673948b808> in <module>
----> 1 pd.read_sql("""SELECT isbn, title, edition, copyright
      2                 FROM titles
      3                 WHERE title NOT LIKE '%How to Program'
      4                 ORDER BY titel""", connection)

~\anaconda3\lib\site-packages\pandas\io\sql.py in read_sql(sql, con, index_col, coerce_float, params, parse_dates, columns, chunksize)
    493 
    494     if isinstance(pandas_sql, SQLiteDatabase):
--> 495         return pandas_sql.read_query(
    496             sql,
    497             index_col=index_col,

~\anaconda3\lib\site-packages\pandas\io\sql.py in read_query(self, sql, index_col, coerce_float, params, parse_dates, chunksize)
   1769 
   1770         args = _convert_params(sql, params)
-> 1771         cursor = self.execute(*args)
   1772         columns = [col_desc[0] for col_desc in cursor.description]
   1773 

~\anaconda3\lib\site-packages\pandas\io\sql.py in execute(self, *args, **kwargs)
   1735 
   1736             ex = DatabaseError(f"Execution failed on sql '{args[0]}': {exc}")
-> 1737             raise ex from exc
   1738 
   1739     @staticmethod

DatabaseError: Execution failed on sql 'SELECT isbn, title, edition, copyright
                FROM titles
                WHERE title NOT LIKE '%How to Program'
                ORDER BY titel': no such column: titel

pd.read_sql("""SELECT isbn, title, edition, copyright
                FROM titles
                WHERE title NOT LIKE '%How to Program'
                ORDER BY title""", connection)
Out[43]: 
         isbn                          title  edition copyright
0  0134289366      Android 6 for Programmers        3      2016
1  0135404673  Intro to Python for CS and DS        1      2020
